common/filesystem.cpp

```diff
@@ -0,0 +1,515 @@
+      if (rmdir_result.ok() || rmdir_result.error().no_entity()) {
+        // Removed here or by something else.
+        continue;
+      }
+      if (rmdir_result.error().not_empty()) {
+        // Found a non-empty directory, add it to our list and continue.
+        dir_entries.push_back(name.str());
+        continue;
+      }
+
+      // Otherwise, some unknown error, so propagate that.
+      return std::move(rmdir_result).error();
+    }
+
+    dir_stack.push_back({*std::move(read_result), dir_entries_start});
+    return Success();
+  };
+
+  CARBON_RETURN_IF_ERROR(push_dir(*this, path));
+
+  while (true) {
+    auto& current = dir_stack.back();
+    if (current.dir_entries_start != static_cast<ssize_t>(dir_entries.size())) {
+      CARBON_CHECK(current.dir_entries_start <
+                   static_cast<ssize_t>(dir_entries.size()));
+      CARBON_RETURN_IF_ERROR(push_dir(current.dir, dir_entries.back()));
+      continue;
+    }
+
+    dir_stack.pop_back();
+    if (dir_stack.empty()) {
+      break;
+    }
+
+    // Pop this entry of the parent and remove it.
+    auto& parent = dir_stack.back();
+    CARBON_CHECK(parent.dir_entries_start <
+                 static_cast<ssize_t>(dir_entries.size()));
+    std::filesystem::path subdir_path = dir_entries.pop_back_val();
+    CARBON_RETURN_IF_ERROR(parent.dir.Rmdir(std::move(subdir_path)));
+  }
+
+  return Rmdir(std::move(path));
+}
+
+auto DirRef::ReadlinkSlow(std::filesystem::path path)
+    -> ErrorOr<std::string, PathError> {
+  // Read directly into a string to avoid allocating two large buffers.
+  std::string large_buffer;
+  // Stat the symlink to get an initial guess at the size.
+  CARBON_ASSIGN_OR_RETURN(FileStatus status, Lstat(path));
+  // We try to use the size from the `lstat` unless it is empty, in which case
+  // we try to use `PATH_MAX` or a constant value. We have a fallback to
+  // dynamically discover an adequate buffer size below that will handle any
+  // inaccuracy.
+  ssize_t buffer_size = status.size();
+  if (buffer_size == 0) {
+#ifdef PATH_MAX
+    buffer_size = PATH_MAX;
+#else
+    buffer_size = 1024;
+#endif
+  }
+  large_buffer.resize(status.size());
+  ssize_t result =
+      readlinkat(dfd_, path.c_str(), large_buffer.data(), large_buffer.size());
+  if (result == -1) {
+    return PathError(errno, "Readlink on '{0}' relative to '{1}'",
+                     std::move(path), dfd_);
+  }
+
+  // Now the really bad fallback case: if there are racing writes to the
+  // symlink, the guessed size may not have been large enough. As a last-ditch
+  // effort, begin doubling (from the next power of two >= PATH_MAX) the length
+  // until it fits. We cap this at 10 MiB to prevent egregious file system
+  // contents (or some bug somewhere) from exhausting memory.
+  constexpr ssize_t MaxSize = 10 << 20;
+  constexpr ssize_t MinSize =
+#ifdef PATH_MAX
+      PATH_MAX
+#else
+      1024
+#endif
+      ;
```

Comment:
Maybe move this earlier (and name `FallbackBufferSize` or something) and use this constant for the `buffer_size` as well, instead of duplicating the constants?